airbyte-integrations/connectors/destination-pinecone/destination_pinecone/indexer.py

```diff
@@ -0,0 +1,122 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import itertools
+from logging import INFO
+import os
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import pinecone
+from airbyte_cdk.models.airbyte_protocol import ConfiguredAirbyteCatalog, AirbyteLogMessage, AirbyteMessage, DestinationSyncMode, Level, Type
+from destination_pinecone.config import PineconeIndexingModel
+from destination_pinecone.document_processor import METADATA_RECORD_ID_FIELD, METADATA_STREAM_FIELD
+from destination_pinecone.embedder import Embedder
+from destination_pinecone.measure_time import measure_time
+from destination_pinecone.utils import format_exception
+from langchain.schema import Document
+
+class Indexer(ABC):
+    def __init__(self, config: Any, embedder: Embedder):
+        self.config = config
+        self.embedder = embedder
+        pass
+
+    def pre_sync(self, catalog: ConfiguredAirbyteCatalog):
+        pass
+
+    def post_sync(self) -> List[AirbyteMessage]:
+        return []
+
+    @abstractmethod
+    def index(self, document_chunks: List[Document], delete_ids: List[str]):
+        pass
+
+    @abstractmethod
+    def check(self) -> Optional[str]:
+        pass
+
+    @property
+    def max_metadata_size(self) -> Optional[int]:
+        return None
+
+
+def chunks(iterable, batch_size):
+    """A helper function to break an iterable into chunks of size batch_size."""
+    it = iter(iterable)
+    chunk = tuple(itertools.islice(it, batch_size))
+    while chunk:
+        yield chunk
+        chunk = tuple(itertools.islice(it, batch_size))
+
+
+# large enough to speed up processing, small enough to not hit pinecone request limits
+PINECONE_BATCH_SIZE = 40
+
+
+class PineconeIndexer(Indexer):
+    config: PineconeIndexingModel
+
+    def __init__(self, config: PineconeIndexingModel, embedder: Embedder):
+        super().__init__(config, embedder)
+        pinecone.init(api_key=config.pinecone_key, environment=config.pinecone_environment, threaded=True)
+        index_description = pinecone.describe_index(config.index)
+        self.pod_type = index_description.pod_type
+        
+        self.pinecone_index = pinecone.Index(config.index, pool_threads=10)
+        self.embed_fn = measure_time(self.embedder.langchain_embeddings.embed_documents)
+        self.embedding_dimensions = embedder.embedding_dimensions
+
+    def pre_sync(self, catalog: ConfiguredAirbyteCatalog):
+        for stream in catalog.streams:
+            if stream.destination_sync_mode == DestinationSyncMode.overwrite:
+                self.delete_vectors(filter={METADATA_STREAM_FIELD: stream.stream.name})
+
+    def post_sync(self):
+        return [AirbyteMessage(type=Type.LOG, log=AirbyteLogMessage(level=Level.WARN, message=self.embed_fn._get_stats()))]
+
+    def delete_vectors(self, filter):        
+        if self.pod_type == "starter":
+            # Starter pod types have a maximum of 1000000 rows
```

Comment:
If there's a 1M rows in a starter pod, this will just delete 10K out of them, right? Is this a problem?

Reply:
There's actually 100,000k, I had the number wrong. This will just have to be part of the limitation. Could we limit the batch size to 10k, such that if 10k entries need to be synced, this will work as intended?

Reply:
Discussed offline, we'll just go with the 10k for now and document it